lib/puppet/ssl/state_machine.rb

```diff
@@ -393,25 +393,42 @@ def ensure_client_certificate
   private
 
   def run_machine(state, stop)
-    with_lock do
-      loop do
-        state = run_step(state)
-
-        case state
-        when stop
-          break
-        when Error
-          if @onetime
-            Puppet.log_exception(state.error)
-            raise state.error
+    begin
+      with_lock do
+        loop do
+          state = run_step(state)
+
+          case state
+          when stop
+            break
+          when Error
+            if @onetime
+              Puppet.log_exception(state.error)
+              raise state.error
+            end
+          else
+            # fall through
           end
-        else
-          # fall through
         end
       end
-    end
 
-    state
+      state
+    rescue Puppet::LockError
+      wait_for_lock_deadline ||= Time.now.to_i + Puppet[:maxwaitforlock]
+
+      if Puppet[:waitforlock] < 1
+        raise Puppet::Error, _('Another puppet instance is already running; exiting')
+      elsif Time.now.to_i > wait_for_lock_deadline
+        raise Puppet::Error, _("Exiting now because the maxwaitforlock timeout has been exceeded.")
+      else
+        while File.file?(@lockfile.file_path) && Time.now.to_i <= wait_for_lock_deadline
+          Puppet.info _("Another puppet instance is already running; --waitforlock flag used, waiting for running instance to finish.")
+          Puppet.info(_("Will try again in %{time} seconds.") % {time: Puppet[:waitforlock]})
+          sleep Puppet[:waitforlock]
+        end
+        retry
+      end
+    end
```

Comment:
I'm going to have to make some non-trivial change to the ssl state machine for a different ticket. I'm thinking it'd be best to not change the state machine here, and I can take care of that as part of the other ticket (PUP-10228).

Reply:
I filed a new ticket to add waitforlock support to the ssl lock https://tickets.puppetlabs.com/browse/PUP-10420